CMakeLists.txt

```diff
@@ -543,14 +549,6 @@ set_target_properties(torrent-rasterbar
 		SOVERSION ${SOVERSION}
 )
 
-if (cxx_std_14 IN_LIST CMAKE_CXX_COMPILE_FEATURES)
-	target_compile_features(torrent-rasterbar PUBLIC cxx_std_14)
-	message(STATUS "Building in C++14 mode")
-else()
-	target_compile_features(torrent-rasterbar PUBLIC cxx_std_11)
-	message(STATUS "Building in C++11 mode")
-endif()
```

Comment:
it's not safe to set the c++ version just for a single target. The c++ version must be set at the top level, not at this library.

Reply:
It was here to propagate this value into the exported targets and the pkg-config file. As a `PUBLIC` feature it will be picked up by all dependent targets. I don't understand therefore why it is unsafe.

Reply:
picking which C++ version to used *just* based on what the compiler supports is really dangerous and fails in lots of ways, especially when linking against pre-built system libraries. Even if this property is propagated up to dependents, will the build fail if two dependencies propagate conflicting properties up?

my understanding of the cmake model is that these kinds of ABI altering options should always only be propagated down the dependency tree, and only from the top level. (and libtorrent, being a library, won't ever really be the top level)

Reply:
I didn't realize this affected the `pkg-cfg` though, but is it important that it does? it seems suspicious to put these options in there too (for the same reason)

Reply:
How can we propagate that down if libtorrent is built already and we are compiling an application using it? libtorrent ABI depends on the C++ version, right? If so, we have to set something for clients, otherwise the library has to be rebuild with compiler upgrade/downgrade which changes the default C++ standard.

> …will the build fail if two dependencies propagate conflicting properties up?

Errr… It was you who coded it in that way to fail linking in such cases. Why do you ask?

Reply:
Or am I wrong here and it fails in run time?

Reply:
you're right, for ``pkg-cfg`` it can't be propagated down, and you may as well propagate it up.
I don't think I've really contributed much to the cmake file. My understanding is that detecting conflicting properties being propagated up is a limitation of cmake, that it can't do that, and that's why the convention is to only propagate down, from top level.

does the package config file look at the build options for the `torrent-rasterbar` target to build the file? if so, this may still be correct. By removing this, the build options are all derived from the top level `cmake` invocation of `CMAKE_CXX_VERSION`, right?

Reply:
how can I verify that the pkg-config file is correct? I don't actually see the `libtorrent-rasterbar.pc` be updated at all by cmake. Is it put in some other place?

Reply:
> how can I verify that the pkg-config file is correct?

After the generation stage (i.e. just before the build starts) it can be found inside `torrent-rasterbar-pkgconfig` subdir in the build directory. 

>  I don't actually see the libtorrent-rasterbar.pc be updated at all by cmake. Is it put in some other place?

Do you mean the install target does not actually install the pkg-config file into the proper location? It should do that (qlist lists contents of package on Gentoo):
```
$ qlist libtorrent-rasterbar | grep pkgconfig
/usr/lib64/pkgconfig/libtorrent-rasterbar.pc
```

Reply:
I see, the default `CXX_STANDARD` is private. I agree that whatever `CXX_STANDARD` resolves to should be put in the pkg-config file. However, pkg-config will most likely fail in the case of a client linking against two separate libraries, both which have conflicting c++ versions exported. But I don't think there's anything that can be done about that in pkg-config. However, in cmake things can be a bit better by never setting this explicitly, but only rely on the target pulling in libtorrent passing down its settings.

do you think it would work to do this unconditionally?

`target_compile_features(torrent-rasterbar PUBLIC cxx_std_${CXX_STANDARD})`

Reply:
> Do you mean the install target does not actually install the pkg-config file into the proper location? It should do that (qlist lists contents of a package on Gentoo):

Oh, I see. The automake build scripts puts the `.pc` file in the root directory (based on the `libtorrent-rasterbar.pc.in` also in the root directory). I thought cmake would write it to the same place.

Reply:
> I thought cmake would write it to the same place.

The file is accompanied by a number of others (intermidiates), as you see. I decided to put them together into a subdir, so the root build dir looks cleaner.

Reply:
> but I would not use the variable directly

Psychological self-analysis showed :) that I expect the property to be affected by other ways, and I think to maintain consistency it is safer to read the final property value instead of relying on one of the way to initiate it.

Reply:
in the first version of this patch I set `CXX_STANDARD_REQUIRED`, but I figured it's better to let it be set at the top level. you're suggesting it should be exported in the `pkg-config` file as well?

Reply:
> you're suggesting it should be exported in the pkg-config file as well?

No. As I understand documentation for the `CXX_STANDARD` property and `CMAKE_CXX_STANDARD` variable, the property can have a value, but it is not enforcing unless `CXX_STANDARD_REQUIRED` is set. From the other hand, compile  features are always enforcing.

I guess we have to test `CMAKE_CXX_COMPILE_FEATURES` before copying the value from the property into the compile feature:
```cmake
if (cxx_std_${_std} IN_LIST CMAKE_CXX_COMPILE_FEATURES)
    target_compile_features(torrent-rasterbar PUBLIC cxx_std_${_std})
endif()
```

Reply:
what does it mean for the `CXX_STANDARD` to not be enforced?
is it not reasonable to require this property to be present? Would it not be present if the top-level build command does not specify `CXX_STANDARD`? If so, it should probably still be set by whatever is the default in the compiler, no?

Reply:
If compiler does not support the requested standard version and  `CXX_STANDARD_REQUIRED` is not set, CMake switches to the latest supported by the compiler. See https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html

Reply:
do you think it would be appropriate to set `CXX_STANDARD_REQUIRED` and `CXX_EXTENSIONS` in the libtorrent CMakeLists.txt? What happens if the top-level project sets conflicting values for those? will it still be guaranteed that every part of the program is built with identical flags?

Reply:
it doesn't seem right to set the `target_compile_features` based on `CMAKE_CXX_COMPILE_FEATURES`. The only thing that should be done (as far as I can tell) is to find the `cxx_std_` property in the `torrent-rasterbar` target, and make it `PUBLIC`. (and it seems really odd that cmake doesn't make it public by default, it's clearly part of the ABI).

The documentation for `CXX_STANDARD` says that it will be set on all targets, so it seems like a fairly safe assumption that it will be set on the `torrent-rasterbar` target. But I suppose there's a risk that the `torrent-rastebar` target has some other, conflicting, property set, in which case there are conflicting PUBLIC and PRIVATE properties. Would cmake fail in this case? (one would hope so).

But maybe it's clearer to explicitly list the properties from the target. Is there a way to do that?